bftengine/tests/simpleStorage/FileStorage.cpp

```diff
@@ -0,0 +1,250 @@
+  if (objectInfo) {
+    write(dataPtr, objectInfo->offset, objectSize, 1,
+          FAILED_TO_WRITE_OBJECT);
+    LOG_INFO(logger_, "FileStorage::handleObjectWrite "
+        << objectInfo->toString());
+    // Update the file metadata with a real object size
+    objectInfo->realSize = objectSize;
+    objectsMetadata_->setObjectInfo(*objectInfo);
+    updateFileObjectMetadata(*objectInfo);
+  }
+}
+
+void FileStorage::handleObjectRead(uint16_t objectId, char *outBufferForObject,
+                                   uint32_t &outActualObjectSize) {
+  MetadataObjectInfo *objectInfo = objectsMetadata_->getObjectInfo(objectId);
+  if (objectInfo) {
+    read(outBufferForObject, objectInfo->offset, objectInfo->realSize, 1,
+         FAILED_TO_READ_OBJECT);
+    outActualObjectSize = objectInfo->realSize;
+    LOG_INFO(logger_, "FileStorage::handleObjectRead "
+        << objectInfo->toString());
+  }
+}
+
+void FileStorage::read(uint16_t objectId, uint32_t bufferSize,
+                       char *outBufferForObject,
+                       uint32_t &outActualObjectSize) {
+  LOG_INFO(logger_, "FileStorage::read objectId="
+      << objectId << ", bufferSize=" << bufferSize);
+  verifyOperation(objectId, bufferSize, outBufferForObject);
+  lock_guard<mutex> lock(ioMutex_);
+  handleObjectRead(objectId, outBufferForObject, outActualObjectSize);
+}
+
+void FileStorage::atomicWrite(uint16_t objectId, char *data,
+                              uint32_t dataLength) {
+  LOG_INFO(logger_, "FileStorage::atomicWrite objectId="
+      << objectId << ", dataLength=" << dataLength);
+  verifyOperation(objectId, dataLength, data);
+  lock_guard<mutex> lock(ioMutex_);
+  handleObjectWrite(objectId, data, dataLength);
+}
+
+void FileStorage::beginAtomicWriteOnlyTransaction() {
+  LOG_INFO(logger_, "FileStorage::beginAtomicWriteOnlyTransaction");
+  verifyFileMetadataSetup();
+  if (transaction_) {
+    LOG_INFO(logger_, "FileStorage::beginAtomicWriteOnlyTransaction "
+                      "Transaction has been opened before; ignoring.");
+    return;
+  }
+  transaction_ = new ObjectIdToRequestMap;
+}
+
+void FileStorage::writeInTransaction(uint16_t objectId, char *data,
+                                     uint32_t dataLength) {
+  LOG_INFO(logger_, "FileStorage::writeInTransaction objectId="
+      << objectId << ", dataLength=" << dataLength);
+  verifyOperation(objectId, dataLength, data);
+  if (!transaction_) {
+    LOG_ERROR(logger_, "FileStorage::writeInTransaction " << WRONG_FLOW);
+    throw runtime_error(WRONG_FLOW);
+  }
+  char *buf = new char[dataLength];
+  memcpy(buf, data, dataLength);
+  transaction_->insert(pair<uint16_t, RequestInfo>
+                           (objectId, RequestInfo(buf, dataLength)));
+}
+
+void FileStorage::commitAtomicWriteOnlyTransaction() {
+  LOG_INFO(logger_, "FileStorage::commitAtomicWriteOnlyTransaction");
+  verifyFileMetadataSetup();
+  if (!transaction_) {
```

Comment:
You need to lock above verifyFileMetadataSetup since `objectsMetadata` and `transaction_` are what you are protecting. I'd just move the lock_guard below to the top of the function.